### PR TITLE
Merge ansibleVars from the role into the node

### PR DESCRIPTION
**Pull request by fultonj**
ansibleVars need to be merged from the role into the node during inheritance.

Closes: [OSP-23184](https://issues.redhat.com//browse/OSP-23184)

### CODE REVIEW

**Comment by fultonj** [comment]
Here are results of testing with these two CRs for role and node.

https://github.com/openstack-k8s-operators/dataplane-operator/blob/main/config/samples/dataplane_v1beta1_openstackdataplanenode_edpm_compute_0.yaml

https://github.com/openstack-k8s-operators/dataplane-operator/blob/main/config/samples/dataplane_v1beta1_openstackdataplanerole_edpm_compute.yaml

```
[fultonj@hamfast edpm]$ oc get configmap dataplanerole-edpm-compute-inventory -o yaml
apiVersion: v1
data:
  inventory: |
    edpm-compute:
        vars:
            ansible_port: 22
            ansible_user: root
            dns_search_domains: []
            edpm_chrony_ntp_servers:
                - clock.redhat.com
                - clock2.redhat.com
            management_network: ctlplane
            neutron_physical_bridge_name: br-ex
            neutron_public_interface_name: eth0
kind: ConfigMap
metadata:
  creationTimestamp: "2023-03-16T18:42:36Z"
  name: dataplanerole-edpm-compute-inventory
  namespace: openstack
  resourceVersion: "2921728"
  uid: f45d24b6-42bc-4b3c-847c-00ae59a2c90b
[fultonj@hamfast edpm]$ 
[fultonj@hamfast edpm]$ oc get configmap dataplanenode-edpm-compute-0-inventory -o yaml
apiVersion: v1
data:
  inventory: |
    all:
        hosts:
            edpm-compute-0:
                ansible_host: 192.168.122.100
                ansible_port: "22"
                ansible_user: root
                dns_search_domains: []
                edpm_chrony_ntp_servers:
                    - clock.redhat.com
                    - clock2.redhat.com
                managed: "false"
                management_network: ctlplane
                network_config: {}
                networks: []
                neutron_physical_bridge_name: br-ex
                neutron_public_interface_name: eth0
                tenant_ip: 192.168.24.100
kind: ConfigMap
metadata:
  creationTimestamp: "2023-03-16T18:43:12Z"
  labels:
    openstackdataplanerole: edpm-compute
  name: dataplanenode-edpm-compute-0-inventory
  namespace: openstack
  resourceVersion: "2922127"
  uid: 3c63464e-ebf8-4002-b8ea-905b52967f92
[fultonj@hamfast edpm]$ 
```

File: controllers/openstackdataplanenode_controller.go [hunk 1]
```diff
@@ -394,12 +394,37 @@ func (r *OpenStackDataPlaneNodeReconciler) GetAnsibleNetworks(instance *dataplan
...
+		role[k] = v
+	}
+
+	// Marshal the merged data back into a string of YAML
+	mergedYaml, err := yaml.Marshal(role)
+	if err != nil {
+		r.Log.Error(err, fmt.Sprintf("Failed to unmarshal merged YAML data from role and node AnsibleVars '%s'", string(mergedYaml)))
+		return ""
+	}
+	return string(mergedYaml)
```
Comment (fao89) [inline controllers/openstackdataplanenode_controller.go:427]: we don't need to Marshal it,
we can remove the Unmarshal here:
https://github.com/openstack-k8s-operators/dataplane-operator/blob/main/controllers/openstackdataplanenode_controller.go#L292

Comment (fultonj) [reply controllers/openstackdataplanenode_controller.go:427]: OK, that will be more efficient.
I was trying to not change the input and output types but the call on line 292 is the only one so it doesn't matter.

Comment (fultonj) [reply controllers/openstackdataplanenode_controller.go]: OK, I've made this update. 

https://github.com/openstack-k8s-operators/dataplane-operator/compare/6c212c9acea21b15d87c71757cf2bc178e73ce6e..333630cf7f4779df3b35e4d2375fd072d458ec1d

File: controllers/openstackdataplanenode_controller.go [hunk 2]
```diff
@@ -394,12 +393,30 @@ func (r *OpenStackDataPlaneNodeReconciler) GetAnsibleNetworks(instance *dataplan
...
-	return instanceRole.Spec.NodeTemplate.AnsibleVars
+	nodeYamlError := yaml.Unmarshal([]byte(instance.Spec.Node.AnsibleVars), &node)
+	if nodeYamlError != nil {
+		r.Log.Error(nodeYamlError, fmt.Sprintf("Failed to unmarshal YAML data from node AnsibleVars '%s'", instance.Spec.Node.AnsibleVars))
+		return nil, nodeYamlError
+	}
+
+	// Merge the two maps
+	for k, v := range node {
+		role[k] = v
```
Comment (slagle) [inline controllers/openstackdataplanenode_controller.go:417]: role could be nil here if ansibleVars was not set on the role. Probably need to check both role and node for nil

Comment (fultonj) [reply controllers/openstackdataplanenode_controller.go:417]: OK, I'll send in a follow up on top of this.

Comment (fultonj) [reply controllers/openstackdataplanenode_controller.go:417]: OK, I think that's addressed by this https://github.com/openstack-k8s-operators/dataplane-operator/pull/99

**Pull request merged**